### ISSUE TITLE
undocumented `.azcopysyncmeta` additional files created

### ISSUE THREAD

**Issue by demoray** [Azure/azure-storage-azcopy#2149]
##### Note: The version is visible when running AzCopy without any argument

10.18.0

### Which platform are you using? (ex: Windows, Mac, Linux)

Linux (Ubuntu 20.04)

##### Note: Please remove the SAS to avoid exposing your credentials. If you cannot remember the exact command, please retrieve it from the beginning of the log file.

`azcopy sync . 'CONTAINER_SAS_URL' --compare-hash md5`

### What problem was encountered?

For every `$FILE` in the local directory being synced, there is a new file created named `.$FILE.azcopysyncmeta`.

Syncing from a local directory to a storage container should not result in files being created in the local directory.  There is no reference to these files being created in the --help or in learn.microsoft.com.  This could have significant impact on users where files are unexpectedly updated (when run on Windows) or additional files created (on non-Windows).

This issue was added with the addition of #2020.

### Have you found a mitigation/solution?

**Comment by adreed-msft** [comment]
Documentation will be added to reference these. This is intended behaviour.

**Comment by demoray** [comment]
Modifying the local files and file systems when uploading should not be the expected behavior.

This will have _significant_ impact on a large number of business systems.

Consider similar tools used elsewhere, this behavior is not common across the industry:
* Open source: [rsync](https://download.samba.org/pub/rsync/rsync.1)
* Amazon: [aws s3 sync](https://awscli.amazonaws.com/v2/documentation/api/latest/reference/s3/sync.html)
* GCP: [gsutil rsync](https://cloud.google.com/storage/docs/gsutil/commands/rsync)
* Backblaze: [b2 sync](https://www.backblaze.com/b2/docs/quick_command_line.html)

Consider the use case that led to Azure Storage implementing [immutable legal holds](https://learn.microsoft.com/en-us/azure/storage/blobs/immutable-legal-hold-overview), where users need data to keep exact copies of files and file systems for legal/forensic purposes.

**Comment by adreed-msft** [comment]
> Modifying the local files and file systems when uploading should not be the expected behavior.
> 
> This will have _significant_ impact on a large number of business systems.
> 
> Consider similar tools used elsewhere, this behavior is not common across the industry:
> 
> * Open source: [rsync](https://download.samba.org/pub/rsync/rsync.1)
> * Amazon: [aws s3 sync](https://awscli.amazonaws.com/v2/documentation/api/latest/reference/s3/sync.html)
> * GCP: [gsutil rsync](https://cloud.google.com/storage/docs/gsutil/commands/rsync)
> * Backblaze: [b2 sync](https://www.backblaze.com/b2/docs/quick_command_line.html)
> 
> Consider the use case that led to Azure Storage implementing [immutable legal holds](https://learn.microsoft.com/en-us/azure/storage/blobs/immutable-legal-hold-overview), where users need data to keep exact copies of files and file systems for legal/forensic purposes.

These files are only created when using --compare-hash and are intended to be a caching of the file's hash at a specific LMT. Because the calculation of MD5s can be slow, we opt to try and store said data for re-use later where possible, assuming that a customer is going to perform an incremental sync.

For Windows, these are stored under an Alternate Data Stream on the file. Technically, this limits this functionality to NTFS. But it is a "cleaner" solution than the hidden files. For other OSes, we opted to simply use the hidden file.

What could be done is we could provide the option to set the cache directory to another location and not use the ADS period, or provide an option to simply disable the cache all together.

**Comment by adreed-msft** [comment]
Do you have any input or ideas for a better solution? Customer feedback is very important to us.

**Comment by alexlod** [comment]
Would be great for the hash to be stored as metadata with the file itself. Not sure if blob storage supports that or not.

Another nice feature would be a simple delete utility to delete all those files. Now I'll need to do a very precarious `find` command...

**Comment by alexlod** [comment]
Here's the find command I ran locally:

```
find . -regextype posix-extended -regex '.*azcopysyncmeta$' -mtime +3 -delete
```

Then I ran sync again with `--delete-destination=true` to delete them from the destination.

**Comment by adreed-msft** [comment]
Sounds like the best action here will be to support using xattr where possible and then supporting a separate targeted directory  where that's not available.

**Comment by adreed-msft** [comment]
https://github.com/Azure/azure-storage-azcopy/pull/2214 adds additional methods for hash storage, namely xattr for MacOS/Linux and ADS on Windows. XAttr and ADS are the defaults, with HiddenFiles being a secondary option including the ability to store said data in another directory entirely.

When utilizing XAttr/ADS, it's expected that the filesystem(s) in which the source would traverse all support the relevant OS-specific features utilized.

**Comment by siminsavani-msft** [comment]
Fixed in 10.19.0!
